### DISCUSSION TITLE
Questions on Building a Queue Worker with Supabase Edge Functions

### DISCUSSION THREAD

**Discussion by jumski** [supabase/supabase#31246] [Questions] [answered]
**Hello everyone,**

I’m currently working on an **open-source queue worker** built on top of **Supabase Edge Functions** as part of a **larger, Postgres-centric workflow orchestration engine** that I’ve been developing full-time for the last two months. I’d greatly appreciate any guidance on leveraging the Edge Runtime to build a robust, community-oriented solution. Insights from both the Supabase team and the community would be invaluable.

---

**Current Approach**  
- The worker process starts when the Edge Function boots (outside the Deno request handler).  
- Immediately after boot, its main loop is scheduled using `waitUntil`.  
- It connects to Postgres and uses `pgmq.read_with_poll` to continuously read messages from a target queue.  
- When a message arrives, the handler runs, and its promise is also managed via `waitUntil`.

---

**Handling CPU/Wall Time Limits**  
At some point, the worker may hit its CPU or wall-clock time limit, triggering the `onbeforeunload` event. From what I’ve learned, once `onbeforeunload` fires, that Edge Function instance no longer accepts new requests. To ensure continuity, I issue an HTTP request to `/api/v1/functions/my-worker-fn`, effectively spawning a new Edge Function instance that starts a fresh worker.

**Disclaimer:**  
I’m not currently tracking these workers in a database nor performing any heartbeats. However, I will add this soon to control the number of spawned workers and better manage the overall process.

---

**Questions**  
1. **Compliance:** Is this pattern—specifically spawning a new instance when `onbeforeunload` fires—aligned with the Terms of Service for Edge Functions?  
2. **Instance Limits:** Are there any rules or limitations on how frequently new instances can be spawned or how many instances can run concurrently?  
3. **Graceful Shutdown:** Beyond `onbeforeunload`, are there other events or APIs I can use to achieve a more graceful shutdown process?  
4. **Roadmap & Collaboration:** Is Supabase considering a built-in solution for similar use cases? I’d be happy to collaborate rather than reinvent the wheel.

---

**Next Steps & Feedback**  
I plan to release this worker component soon to gather feedback. While it’s just one building block of the larger orchestration engine I’m developing, ensuring it aligns with best practices and community standards is a top priority.

Thank you for your time and any insights you can share!

**Comment by GaryAustin1** [comment, marked as answer]
You might want to post this in supabase/edge-runtime github.  You are more likely to get a edge dev to see it there.

**Reply by jumski** [reply]
Yup I will do that, landed here because github advised me to use discussion for the questions.

I will just create an issue in the rwpo itself. Thanks!

**Comment by tmcgann** [comment]
If you want to follow the convo: https://github.com/supabase/edge-runtime/issues/464
